README.md
[![Version](https://img.shields.io/badge/pod-2.0-green.svg)](https://cocoapods.org/pods/MultiStepSlider)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](http://cocoadocs.org/docsets/MultiStepSlider)
[![Documents](https://img.shields.io/badge/platform-iOS-orange.svg?style=flat)](http://cocoadocs.org/docsets/MultiStepSlider)

# MultiStepSlider

A custom UIControl which functions like UISlider where you can set multiple intervals with different step values for each interval. This is useful when an interval spans over large values, for example, price of real estates. In that case it is convenient to divide the large interval in smaller intervals with each interval having its own step value.

#Installation

Add following lines in your pod file if you are using Swift 3
```
pod ‘MultiStepSlider’, '~> 2.0'

```
 
Add following lines in your pod file for previous Swift versions
```
pod ‘MultiStepSlider’, '~> 1.4'
```
#Usage

```objc
import ‘MultiStepSlider’
```

In interface builder, drag one UIView and set its class as MultiStepSlider.

![](https://cloud.githubusercontent.com/assets/3590619/16220300/a3d884ea-37a9-11e6-8732-76d16422ba57.png)

#### Configuration

**MultiStepSlider** can be configured by the following method. 
```
func configureSlider(intervals intervals: [Interval], preSelectedRange: RangeValue?)
```
The first parameter is an array of type **Interval** which is defined as:
```
 public struct Interval {
	public private(set) var min: Float = 0.0
	public private(set) var max: Float = 0.0
	public private(set) var stepValue: Float = 0.0
 }
```
The second parameter is of type **RangeValue** which is defined as:
```
 public struct RangeValue {
	public var lower: Float = 0.0
	public var upper: Float = 0.0
 }
```
This dictates the initial positions for lower and upper thumb. The _lower_ and _upper_ of **RangeValue** should lie within the interval specified and should be a valid node value. For example, if the there is an interval
**Interval(min: 50000, max: 100000, stepValue: 10000)**, then **60000** will be a valid node, but not **65000**. In that case, a warning will be shown.

```
Warning: Range contains invalid node
```

#### Example
```
@IBOutlet weak var slider: MultiStepRangeSlider!
override func viewDidLoad() {
super.viewDidLoad()
let intervals = [Interval(min: 50000, max: 100000, stepValue: 10000),
Interval(min: 100000, max: 1000000, stepValue: 100000),
Interval(min: 1000000, max: 3000000, stepValue: 500000)]
let preSelectedRange = RangeValue(lower: 80000, upper: 500000)
slider.configureSlider(intervals: intervals, preSelectedRange: preSelectedRange)
print("continuous: lower = \(slider.continuousCurrentValues.lower) higher = \(slider.continuousCurrentValues.upper)")
print("discrete: lower = \(slider.discreteCurrentValue.lower) higher = \(slider.discreteCurrentValue.upper)")
}
```
<img src="https://cloud.githubusercontent.com/assets/3590619/16224574/4260d8a4-37c0-11e6-8d39-7d9c6b7497af.gif" width="400" display="inline-block">

# Properties

#### discreteCurrentValue

This is of type RangeValue and gives the discrete upper and lower value.

#### continuousCurrentValue

This is of type RangeValue and gives the continuous upper and lower value.

#### trackTintColor

This color is used to tint the part of the track which is outside the range of lower thumb and upper thumb.

The default color is lightGrayColor.

#### trackHighlightTintColor

The color used to tint the part of the track which is inside the range of lowerValue and upperValue.

The default color is #007AFF (rgba = 0, 122, 255, 1)

#### trackCurvaceousness

This property is used to control the curvature of ends of the track. The property can have value from 0 to 1.
<p>trackCurvaceousness = 0.0</p><span>
<img src="https://cloud.githubusercontent.com/assets/3590619/16221343/dd8df9e4-37af-11e6-93d8-626cb9d4aedb.png"  width="200" height="30" display="inline-block"></span>
<p>trackCurvaceousness = 1.0</p><span>
<img src="https://cloud.githubusercontent.com/assets/3590619/16221294/9e9149f8-37af-11e6-9f97-b088fdb90869.png"  width="200" height="40" display="inline-block"/></span>

#### thumbCurvaceousness

This property is used to control the curvature of the thumbs. The property can have value from 0 to 1. 

<p>trackCurvaceousness = 0.5</p><span>
<img src="https://cloud.githubusercontent.com/assets/3590619/16221988/599a7910-37b3-11e6-8f3d-8dc293dde98f.png"  width="200" height="30" display="inline-block"></span>
<p>trackCurvaceousness = 1.0</p><span>
<img src="https://cloud.githubusercontent.com/assets/3590619/16221294/9e9149f8-37af-11e6-9f97-b088fdb90869.png"  width="200" height="40" display="inline-block"/></span>

#### shadowEnabled

Setting this property as true will show shadow around the thumbs.

# License

MultiStepSlider is available under the MIT License.

# Reference

https://github.com/warchimede/RangeSlider/


CHANGELOG.md
version 2.0

* Update for Swift 3.

version 1.4

* Update access controls for variables. Now the properties `discreteCurrentValue`, `continuousCurrentValue`, and the properties of struct `Interval` and `Range` are accessible, but are read only.

version 1.3

* IBInspectable properties made public

version 1.2

* Change the class name from MultiStepRangeSlider to MultiStepSlider
* Add documentation

version 1.1

* First version


MultiStepSlider/AppDelegate.swift
//
//  AppDelegate.swift
//  MultiStepSlider
//
//  Created by Susmita Horrow on 11/01/16.
//  Copyright © 2016 hsusmita. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

	var window: UIWindow?


	func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
		// Override point for customization after application launch.
		return true
	}

	func applicationWillResignActive(_ application: UIApplication) {
		// Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
		// Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
	}

	func applicationDidEnterBackground(_ application: UIApplication) {
		// Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
		// If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
	}

	func applicationWillEnterForeground(_ application: UIApplication) {
		// Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
	}

	func applicationDidBecomeActive(_ application: UIApplication) {
		// Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
	}

	func applicationWillTerminate(_ application: UIApplication) {
		// Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
	}


}



MultiStepSlider/Source/Utility.swift
//
//  Utility.swift
//  MultiStepSlider
//
//  Created by Susmita Horrow on 31/10/16.
//  Copyright © 2016 hsusmita. All rights reserved.
//

import Foundation
import UIKit

public struct RangeValue {
	public var lower: Float = 0.0
	public var upper: Float = 0.0
	
	public init(lower: Float, upper: Float) {
		self.lower = lower
		self.upper = upper
	}
}

/**
The struct to encapsulate the information of interval.

Initialization:
```
Interval(min: 50000, max: 100000, stepValue: 10000)
```
*/

public struct Interval {
	public private(set) var min: Float = 0.0
	public private(set) var max: Float = 0.0
	public private(set) var stepValue: Float = 0.0
	public private(set) var nodes: [Float] = []
	
	public init(min: Float, max: Float, stepValue: Float) {
		self.min = min
		self.max = max
		self.stepValue = stepValue
		var index = min
		while index <= max {
			self.nodes.append(index)
			index += stepValue
		}
	}
}



MultiStepSlider/ViewController.swift
//
//  ViewController.swift
//  MultiStepSlider
//
//  Created by Susmita Horrow on 11/01/16.
//  Copyright © 2016 hsusmita. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

	@IBOutlet weak var minimumLabel: UILabel!
	@IBOutlet weak var maximimLabel: UILabel!
	@IBOutlet weak var slider: MultiStepRangeSlider!
	let numberFomatter = NumberFormatter()

	override func viewDidLoad() {
		super.viewDidLoad()
		let intervals = [Interval(min: 50000, max: 100000, stepValue: 10000),
						Interval(min: 100000, max: 1000000, stepValue: 100000),
						Interval(min: 1000000, max: 3000000, stepValue: 500000)]
		let preSelectedRange = RangeValue(lower: 80000, upper: 500000)
		slider.configureSlider(intervals: intervals, preSelectedRange: preSelectedRange)
		minimumLabel.text = abbreviateNumber(NSNumber(value: slider.discreteCurrentValue.lower as Float)) as String
		maximimLabel.text = abbreviateNumber(NSNumber(value: slider.discreteCurrentValue.upper as Float)) as String
	}

	@IBAction func handleSliderChange(_ sender: AnyObject) {
		minimumLabel.text = abbreviateNumber(NSNumber(value: slider.discreteCurrentValue.lower as Float)) as String
		maximimLabel.text = abbreviateNumber(NSNumber(value: slider.discreteCurrentValue.upper as Float)) as String
		print("lower = \(slider.continuousCurrentValue.lower) higher = \(slider.continuousCurrentValue.upper)")
	}
}

//http://stackoverflow.com/questions/18267211/ios-convert-large-numbers-to-smaller-format

func abbreviateNumber(_ num: NSNumber) -> NSString {
	var ret: NSString = ""
	let abbrve: [String] = ["K", "M", "B"]

	let floatNum = num.floatValue

	if floatNum > 1000 {

		for i in 0..<abbrve.count {
			let size = pow(10.0, (Float(i) + 1.0) * 3.0)
			if (size <= floatNum) {
				let num = floatNum / size
				let str = floatToString(num)
				ret = NSString(format: "%@%@", str, abbrve[i])
			}
		}
	} else {
		ret = NSString(format: "%d", Int(floatNum))
	}

	return ret
}

func floatToString(_ val: Float) -> NSString {
	var ret = NSString(format: "%.1f", val)
	var c = ret.character(at: ret.length - 1)

	while c == 48 {
		ret = ret.substring(to: ret.length - 1) as NSString
		c = ret.character(at: ret.length - 1)


		if (c == 46) {
			ret = ret.substring(to: ret.length - 1) as NSString
		}
	}
	return ret
}





MultiStepSliderTests/MultiStepSliderTests.swift
//
//  MultiStepSliderTests.swift
//  MultiStepSliderTests
//
//  Created by Susmita Horrow on 11/01/16.
//  Copyright © 2016 hsusmita. All rights reserved.
//

import XCTest
@testable import MultiStepSlider

class MultiStepSliderTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


MultiStepSliderUITests/MultiStepSliderUITests.swift
//
//  MultiStepSliderUITests.swift
//  MultiStepSliderUITests
//
//  Created by Susmita Horrow on 11/01/16.
//  Copyright © 2016 hsusmita. All rights reserved.
//

import XCTest

class MultiStepSliderUITests: XCTestCase {
        
    override func setUp() {
        super.setUp()
        
        // Put setup code here. This method is called before the invocation of each test method in the class.
        
        // In UI tests it is usually best to stop immediately when a failure occurs.
        continueAfterFailure = false
        // UI tests must launch the application that they test. Doing this in setup will make sure it happens for each test method.
        XCUIApplication().launch()

        // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // Use recording to get started writing UI tests.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
}


MultiStepSlider/Source/MultiStepSlider.swift
//
//  MultiStepSlider.swift
//  MultiStepSlider
//
//  Created by Susmita Horrow on 11/01/16.
//  Copyright © 2016 hsusmita. All rights reserved.

import Foundation
import UIKit
import QuartzCore

struct Path {
	var origin: CGFloat = 0
	var length: CGFloat = 0
}

class RangeSliderTrackLayer: CALayer {
	static let defaultHighlightTintColor = UIColor(red: 0, green: 122/255.0, blue: 255.0/255.0, alpha: 1.0)
	static let defaultTintColor = UIColor.lightGray

	var highlightedPath = Path(origin: 0, length: 0)
	var highlightTintColor: UIColor?
	var tintColor: UIColor?
	var curvaceousness: CGFloat = 2.0

	override func draw(in ctx: CGContext) {
		// Clip
		let cornerRadius = bounds.height * curvaceousness / 2.0
		let path = UIBezierPath(roundedRect: bounds, cornerRadius: cornerRadius)
		ctx.addPath(path.cgPath)

		// Fill the track
		ctx.setFillColor((tintColor ?? RangeSliderTrackLayer.defaultTintColor).cgColor)
		ctx.addPath(path.cgPath)
		ctx.fillPath()

		// Fill the highlighted range
		ctx.setFillColor((highlightTintColor ?? RangeSliderTrackLayer.defaultHighlightTintColor).cgColor)
		let rect = CGRect(x: highlightedPath.origin, y: 0.0, width: highlightedPath.length, height: bounds.height)
		ctx.fill(rect)
	}
}

class RangeSliderThumbLayer: CALayer {
	static let defaultTintColor = UIColor(red: 0, green: 122/255.0, blue: 255.0/255.0, alpha: 1.0)
	var highlighted: Bool = false {
		didSet {
			setNeedsDisplay()
		}
	}
	var curvaceousness: CGFloat = 2.0
	var tintColor = defaultTintColor

	override func draw(in ctx: CGContext) {
		let thumbFrame = bounds.insetBy(dx: 2.0, dy: 2.0)
		let cornerRadius = thumbFrame.height * curvaceousness / 2.0
		let thumbPath = UIBezierPath(roundedRect: thumbFrame, cornerRadius: cornerRadius)
		
		// Fill
		ctx.setFillColor(tintColor.cgColor)
		ctx.addPath(thumbPath.cgPath)
		ctx.fillPath()

		// Outline
		let strokeColor = UIColor.clear
		ctx.setStrokeColor(strokeColor.cgColor)
		ctx.setLineWidth(0.5)
		ctx.addPath(thumbPath.cgPath)
		ctx.strokePath()

		// Highlight
		if highlighted {
			ctx.setFillColor(UIColor(white: 0.0, alpha: 0.1).cgColor)
			ctx.addPath(thumbPath.cgPath)
			ctx.fillPath()
		}
	}
}

@IBDesignable

open class MultiStepRangeSlider: UIControl {
	
	/**
	The height of track.
	
	The track is the horizontal line on which the thumbs slide
	*/

	@IBInspectable open var trackLayerHeight: CGFloat = 1.0 {
		didSet {
			var trackFrame = trackLayer.frame
			trackFrame.origin.y = (bounds.size.height - trackLayerHeight)/2
			trackFrame.size.height = trackLayerHeight
			trackLayer.frame = trackFrame
			trackLayer.setNeedsDisplay()
		}
	}
	
	/**
	The color used to tint the part of the track which is outside the range of lowerValue and upperValue.
	
	The default color is lightGrayColor.
	*/

	@IBInspectable open var trackTintColor: UIColor = RangeSliderTrackLayer.defaultTintColor {
		didSet {
			trackLayer.tintColor = trackTintColor
			trackLayer.setNeedsDisplay()
		}
	}
	
	/**
	The color used to tint the part of the track which is inside the range of lowerValue and upperValue.
	
	The default color is #007AFF (rgba = 0, 122, 255, 1)
	*/
	@IBInspectable open var trackHighlightTintColor: UIColor = RangeSliderTrackLayer.defaultHighlightTintColor {
		didSet {
			trackLayer.highlightTintColor = trackHighlightTintColor
			trackLayer.setNeedsDisplay()
		}
	}
	
	/**
	This property is used to control the curvature of ends of the track.
	
	The property can have value from 0 to 1.
	*/
	@IBInspectable open var trackCurvaceousness: CGFloat = 1.0 {
		didSet {
			if trackCurvaceousness < 0.0 {
				trackCurvaceousness = 0.0
			}
			
			if trackCurvaceousness > 1.0 {
				trackCurvaceousness = 1.0
			}
			trackLayer.curvaceousness = trackCurvaceousness
			trackLayer.setNeedsDisplay()
		}
	}
	
	/**
	The size of thumb.
	
	The thumbs mark the lower and upper end of the selected range on the slider
	*/
	@IBInspectable open var thumbSize: CGSize = CGSize(width: 10.0, height: 10.0) {
		didSet {
			updateLayerFrames()
		}
	}
	
	/**
	The color used as tint color for the thumbs
	
	The thumbs mark the lower and upper end of the selected range on the slider
	The default color is #007AFF (rgba = 0, 122, 255, 1)
	*/
	@IBInspectable open var thumbTintColor: UIColor = UIColor.white {
		didSet {
			lowerThumbLayer.tintColor = thumbTintColor
			upperThumbLayer.tintColor = thumbTintColor
			lowerThumbLayer.setNeedsDisplay()
			upperThumbLayer.setNeedsDisplay()
		}
	}
	
	/**
	This property is used to control the curvature of the thumbs.
	
	The property can have value from 0 to 1.
	*/
	@IBInspectable open var thumbCurvaceousness: CGFloat = 1.0 {
		didSet {
			if thumbCurvaceousness < 0.0 {
				thumbCurvaceousness = 0.0
			}
			
			if thumbCurvaceousness > 1.0 {
				thumbCurvaceousness = 1.0
			}
			lowerThumbLayer.curvaceousness = thumbCurvaceousness
			upperThumbLayer.curvaceousness = thumbCurvaceousness
			lowerThumbLayer.setNeedsDisplay()
			upperThumbLayer.setNeedsDisplay()
		}
	}
	
	/**
	Setting this propery adds shadow to both of the thumbs.
	*/
	@IBInspectable open var shadowEnabled: Bool = true {
		didSet {
			if shadowEnabled {
				addShadow(lowerThumbLayer)
				addShadow(upperThumbLayer)
			} else {
				removeShadow(lowerThumbLayer)
				removeShadow(upperThumbLayer)
			}
		}
	}
	
	/**
	This represents the discrete upper and lower values.
	*/
	open var discreteCurrentValue: RangeValue = RangeValue(lower: 0, upper: 1) {
		didSet {
			updateLayerFrames()
		}
	}
	
	/**
	This gives the continuous upper and lower values.
	*/
	open var continuousCurrentValue: RangeValue {
		return RangeValue(lower: lowerValue, upper: upperValue)
	}
	
	// MARK: - Private variables

	private var intervals: [Interval] = []
	private var preSelectedRange: RangeValue?
	private var nodesList = [Float]()
	private var previousLocation = CGPoint()
	private let trackLayer = RangeSliderTrackLayer()
	private let lowerThumbLayer = RangeSliderThumbLayer()
	private let upperThumbLayer = RangeSliderThumbLayer()
	private var lowerCenter: CGFloat = 0.0
	private var upperCenter: CGFloat = 0.0
	private var lowerValue: Float = 0.0 {
		didSet {
			updateLayerFrames()
		}
	}

	private var upperValue: Float = 1.0 {
		didSet {
			updateLayerFrames()
		}
	}

	private var trackFrame: CGRect = CGRect.zero {
		didSet {
			if oldValue.equalTo(trackFrame) {
				return
			}
			trackLayer.frame = trackFrame
			lowerCenter = positionForNodeValue(discreteCurrentValue.lower) ?? trackFrame.minX
			upperCenter = positionForNodeValue(discreteCurrentValue.upper) ?? trackFrame.maxX
			updateLayerFrames()
		}
	}
	
	// MARK: - Life cycle
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		initializeLayers()
	}
	
	public required init?(coder: NSCoder) {
		super.init(coder: coder)
		initializeLayers()
	}

	override open func layoutSubviews() {
		super.layoutSubviews()
		trackFrame = CGRect(x: thumbSize.width/2, y: (bounds.size.height - trackLayerHeight)/2,
		                        width: bounds.size.width - thumbSize.width, height: trackLayerHeight)
	}
	
	// MARK: - Public methods
	
	/**
	The method configures the required variables
	
	- parameter intervals: The array intervals into which the slider is divided.
	preSelectedRange: This dictates the initial positions for lower and upper thumb.
	The lower and upper of RangeValue should lie within the interval specified and should be a valid node value.
	For example, if the there is an interval Interval(min: 50000, max: 100000, stepValue: 10000),
	then 60000 will be a valid node, but not 65000. In that case, a warning will be shown.
	
	*/
	open func configureSlider(intervals: [Interval], preSelectedRange: RangeValue?) {
		self.intervals = intervals
		self.preSelectedRange = preSelectedRange
		updateNodesList()
		if nodesList.count > 0 {
			if let rangeValue = preSelectedRange {
				guard let _ = positionForNodeValue(rangeValue.lower), let _ = positionForNodeValue(rangeValue.upper) else {
					print("Warning: Range contains invalid node")
					return
				}
				discreteCurrentValue = rangeValue
			} else {
				discreteCurrentValue = RangeValue(lower: Float(nodesList[0]), upper: Float(nodesList[nodesList.count-1]))
			}
			lowerValue = discreteCurrentValue.lower
			upperValue = discreteCurrentValue.upper
			lowerCenter = positionForNodeValue(discreteCurrentValue.lower) ?? trackFrame.minX
			upperCenter = positionForNodeValue(discreteCurrentValue.upper) ?? trackFrame.maxX
		}
		updateLayerFrames()
	}
	
	// MARK: - Private methods
	
	// MARK: - Drawing
	private func initializeLayers() {
		trackLayer.contentsScale = UIScreen.main.scale
		layer.addSublayer(trackLayer)

		lowerThumbLayer.contentsScale = UIScreen.main.scale
		layer.addSublayer(lowerThumbLayer)

		upperThumbLayer.contentsScale = UIScreen.main.scale
		layer.addSublayer(upperThumbLayer)
	}

	private func updateLayerFrames() {
		CATransaction.begin()
		CATransaction.setDisableActions(true)
		
		trackLayer.highlightedPath = Path(origin: lowerCenter - thumbSize.width/2,length: upperCenter - lowerCenter)
		trackLayer.setNeedsDisplay()
		
		lowerThumbLayer.frame = CGRect(x: lowerCenter - thumbSize.width/2, y: (bounds.size.height - thumbSize.height)/2,
		                               width: thumbSize.width, height: thumbSize.height)
		lowerThumbLayer.setNeedsDisplay()
		
		upperThumbLayer.frame = CGRect(x: upperCenter - thumbSize.width/2, y: (bounds.size.height - thumbSize.height)/2,
		                               width: thumbSize.width, height: thumbSize.height)
		upperThumbLayer.setNeedsDisplay()
		
		CATransaction.commit()
	}
	
	private func addShadow(_ layer: CALayer) {
		layer.shadowColor = UIColor.black.cgColor
		layer.shadowOffset = CGSize(width: 0.0, height: 2.0)
		layer.shadowOpacity = 0.3
		layer.shadowRadius = 2
		layer.shadowPath = UIBezierPath(roundedRect: layer.bounds.insetBy(dx: 1, dy: 2), cornerRadius: layer.bounds.height/2).cgPath
	}

	private func removeShadow(_ layer: CALayer) {
		layer.shadowOpacity = 0.0
	}
	
	// MARK: - Calculations
	/**
	The method generates node based on the intervals given.
	*/
	
	private func updateNodesList() {
		nodesList.removeAll()
		nodesList = intervals.map { $0.nodes }.flatMap { $0 }
		nodesList.uniqueInPlace()
	}
	
	/**
	The track is divided into number of nodes. Each node is assigned one value.
	
	This method returns x-cordinate of node corresponding to the given value
	
	- parameter value: The node value for which x-cordinate to be calculated
	
	- returns:
	x-cordinate of the node for the given value if such node exist, else returns nil
	
	*/
	private func positionForNodeValue(_ value: Float)-> CGFloat? {
		let scale = Float(trackLayer.frame.size.width) / (Float(nodesList.count))
		var nodeNumber = -1
		for i in 0..<nodesList.count {
			if value == nodesList[i] {
				nodeNumber = i
			}
		}
		guard nodeNumber >= 0 && nodeNumber < nodesList.count else {
			return nil
		}
		let scaledPosition = scale * Float(nodeNumber)
		if nodeNumber == nodesList.count - 1 {
			return trackLayer.frame.maxX
		}else if nodeNumber == 0 {
			return trackLayer.frame.minX
		}else {
			return CGFloat(scaledPosition + Float(trackLayer.frame.minX) + scale/2)
		}
	}
	
	/**
	The track is divided into number of nodes. Each node is assigned one value.
	
	This method fetches the node nearest to the given thumb position and returns the value assinged to the node.
	
	- parameter position: The thumb position for which node value to be calculated
	
	- returns:
	The value assinged to the node if any valid node exists at the given position, else it returns nil.
	*/
	private func nodeValueForPosition(_ position: CGFloat)-> Float? {
		let scale = trackLayer.frame.size.width / CGFloat(nodesList.count)
		let scaledPosition = position - trackLayer.frame.minX
		let nodeNumber = Int(floor(scaledPosition / scale))
		guard nodeNumber < nodesList.count && nodeNumber >= 0 else {
			return nil
		}
		return nodesList[nodeNumber]
	}
	
	/**
	This method returns the actual value corresponding to the thumb position
	
	- parameter position: The position for which value to be calculated
	
	- returns:
	The actual value corresponding to the thumb position if the position is valid, else it returns nil.
	*/
	private func actualValueForPosition(_ position: CGFloat)-> Float? {
		let scale = trackLayer.frame.size.width / CGFloat(nodesList.count)
		let scaledPosition = position - trackLayer.frame.minX
		let nodeNumber = Int(floor(scaledPosition / scale))
		guard nodeNumber < nodesList.count - 1 && nodeNumber >= 0 else {
			return nil
		}
		let nextNode = nodeNumber + 1
		let valueDifference = nodesList[nextNode] - nodesList[nodeNumber]
		let lower = CGFloat(nodeNumber) * scale
		let value = Float(scaledPosition - lower) * valueDifference / Float(scale)
		return nodesList[nodeNumber] + value
	}
	
	/**
	This method updates upperCenter(the center of upper thumb), upperValue and upper property of discreteCurrentValue
	
	- Parameter offset : The distance moved by upper thumb with respect to last location
	
	*/
	private func updateUpperValue(_ offset: CGFloat) {
		upperCenter = boundValue(upperCenter + offset, lowerValue: lowerThumbLayer.frame.midX + thumbSize.width,
		                         upperValue: trackLayer.frame.maxX)
		if let nodeValue = nodeValueForPosition(upperCenter) {
			discreteCurrentValue.upper = nodeValue
		}
		if let actualValue = actualValueForPosition(upperCenter) {
			upperValue = actualValue
		}
	}
	
	/**
	This method updates lowerCenter(the center of lower thumb), lowerValue and upper property of discreteCurrentValue
	
	- Parameter offset : The distance moved by lower thumb with respect to last location
	*/
	private func updateLowerValue(_ offset: CGFloat) {
		lowerCenter = boundValue(lowerCenter + offset,lowerValue: trackLayer.frame.minX,
		                         upperValue:  upperThumbLayer.frame.midX - thumbSize.width)
		if let nodeValue = nodeValueForPosition(lowerCenter) {
			discreteCurrentValue.lower = nodeValue
		}
		if let actualValue = actualValueForPosition(lowerCenter) {
			lowerValue = actualValue
		}
	}
	
	/**
	This method makes sure the given value lies between given upper and lower limit.
	
	- Parameter value: The value to be checked
	lowerValue: The lower limit
	upperValue: The upper limit
	
	
	- returns
	The given value if it lies within the range
	
	The lowerValue if the given value is less than the lowerValue
	
	The upperValue if the given value is greater than the upperValue
	*/
	private func boundValue(_ value: CGFloat, lowerValue: CGFloat, upperValue: CGFloat) -> CGFloat {
		return min(max(value, lowerValue), upperValue)
	}
	
	// MARK: - Touches
	
	open override func beginTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
		previousLocation = touch.location(in: self)
		if lowerThumbLayer.frame.contains(previousLocation) {
			lowerThumbLayer.highlighted = true
		} else if upperThumbLayer.frame.contains(previousLocation) {
			upperThumbLayer.highlighted = true
		}
		return lowerThumbLayer.highlighted || upperThumbLayer.highlighted
	}

	open override func continueTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
		let location = touch.location(in: self)
		let deltaLocation = location.x - previousLocation.x
		previousLocation = location
		
		if lowerThumbLayer.highlighted {
			updateLowerValue(deltaLocation)
			
		} else if upperThumbLayer.highlighted {
			updateUpperValue(deltaLocation)
		}
		updateLayerFrames()
		sendActions(for: .valueChanged)
		return true
	}
	
	override open func endTracking(_ touch: UITouch?, with event: UIEvent?) {
		lowerThumbLayer.highlighted = false
		upperThumbLayer.highlighted = false
	}
}
